src/Trie.mo

```diff
@@ -354,7 +353,48 @@ module {
   /// Purely-functional representation permits _O(1)_ copy, via persistent sharing.
   public func clone<K, V>(t : Trie<K, V>) : Trie<K, V> = t;
 
-  /// Replace the given key's value option with the given one, returning the previous one
+  /// Combine two nodes that may have a reduced size after an entry deletion.
+  func combineReducedNodes<K, V>(left : Trie<K, V>, right : Trie<K, V>) : Trie<K, V> {
+    switch ((left, right)) {
+      case (#empty, #empty) {
+        #empty
+      };
+      case (#leaf(leftLeaf), #empty) {
+        #leaf(leftLeaf)
+      };
+      case (#empty, #leaf(rightLeaf)) {
+        #leaf(rightLeaf)
+      };
+      case ((#leaf(leftLeaf), #leaf(rightLeaf))) {
+        let size = leftLeaf.size + rightLeaf.size;
+        if (size <= MAX_LEAF_SIZE) {
+          let union = List.append(leftLeaf.keyvals, rightLeaf.keyvals);
+          #leaf({ size = size; keyvals = union })
+        } else {
+          branch((left, right))
+        }
+      };
+      case ((left, right)) {
+        branch(left, right)
+      }
+    }
+  };
+
+  /// Replace the given key's value option with the given value, returning the modified trie.
+  /// Also returns the replaced value if the key existed and `null` otherwise.
+  /// Compares keys using the provided function `k_eq`.
+  ///
+  /// Note: Replacing a key's value by `null` removes the key and also shrinks the trie.
+  ///
+  /// For a more detailed overview of how to use a `Trie`,
+  /// see the [User's Overview](#overview).
+  ///
+  /// Example:
+  /// ```motoko include=initialize
+  /// trie := Trie.put(trie, key "test", Text.equal, 1).0;
+  /// trie := Trie.replace(trie, key "test", Text.equal, 42).0;
+  /// assert (Trie.get(trie, key "hello", Text.equal) == ?42);
+  /// ```
   public func replace<K, V>(t : Trie<K, V>, k : Key<K>, k_eq : (K, K) -> Bool, v : ?V) : (Trie<K, V>, ?V) {
```

Comment:
I wonder if this would produce a little less garbage by communicating the previous value using private state and having rec simply return a trie, not a pair, as in RedBlackTree.insert/replace (or what ever it is called). Maybe not worth it.

Reply:
Thanks for the suggestion. I implemented this.